### PR TITLE
[FSTORE-1036] Add usage logging

### PR DESCRIPTION
**Pull request by kennethmhc**
This PR adds/fixes/changes...
- please summarize your changes to the code 
- and make sure to include all changes to user-facing APIs

JIRA Issue: -

Priority for Review: -

Related PRs: -

**How Has This Been Tested?**

- [ ] Unit Tests
- [ ] Integration Tests
- [ ] Manual Tests on VM

**Checklist For The Assigned Reviewer:**

```
- [ ] Checked if merge conflicts with master exist
- [ ] Checked if stylechecks for Java and Python pass
- [ ] Checked if all docstrings were added and/or updated appropriately
- [ ] Ran spellcheck on docstring
- [ ] Checked if guides & concepts need to be updated
- [ ] Checked if naming conventions for parameters and variables were followed
- [ ] Checked if private methods are properly declared and used
- [ ] Checked if hard-to-understand areas of code are commented
- [ ] Checked if tests are effective
- [ ] Built and deployed changes on dev VM and tested manually
- [x] (Checked if all type annotations were added and/or updated appropriately)
```

### CODE REVIEW

File: python/hsfs/usage.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+    if not _is_enabled:
+        return func
+
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        # Disable usage AFTER import hsfs, return function itself
+        if not _is_enabled:
+            return func(*args, **kwargs)
+
+        start_time = time.time()
```
Comment (javierdlrm) [inline python/hsfs/usage.py:199]: In Python 3.3, perf_counter() was introduced which is a bit more accurate ([docs](https://docs.python.org/3/library/time.html#time.perf_counter)). Can we use that one instead?

File: python/hsfs/usage.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+        except Exception as e:
+            exception = e
+            raise e
+        finally:
+            try:
+                end_time = time.time()
+                execution_time = end_time - start_time
+                _method_counter.add(func)
+                # Send log to REST API server
+                if exception or _method_counter.should_sample(func):
```
Comment (javierdlrm) [inline python/hsfs/usage.py:214]: If I understand correctly, the first 100 calls to annotated methods will be logged individually (one request to our logging service per method call), right?
Have you thought/ test/ notice possible performance impact after adding the ThreadPoolExecutor?

File: python/hsfs/usage.py [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+    def _get_method_name(self, m):
+        return m.__module__ + m.__name__
+
+
+logging.basicConfig(stream=sys.__stdout__)
+_logger = logging.getLogger(__name__)
+_logger.setLevel(logging.INFO)
+_conn = http.client.HTTPSConnection("usage.hops.works")
+_executor = concurrent.futures.ThreadPoolExecutor(max_workers=2)
```
Comment (javierdlrm) [inline python/hsfs/usage.py:119]: Can we make `max_workers` configurable as well? Maybe through another environment variable?

File: python/hsfs/usage.py (hunk 2 above)
Comment (kennethmhc) [reply python/hsfs/usage.py:214]: > If I understand correctly, the first 100 calls to annotated methods will be logged individually (one request to our logging service per method call), right?

yes

> Have you thought/ test/ notice possible performance impact after adding the ThreadPoolExecutor?

When logging is off, there is no apparent change in the latency. When it is on, there is around 4ms increase in the latency. In production setting, I would expect user to turn off the usage logging.

Comment (javierdlrm) [reply python/hsfs/usage.py:215]: Alright, then we just need to make sure it's well documented especially for get_feature_vector(s) method.
We could highlight it in the docs with one of those `info` captions

**Review by javierdlrm — APPROVED** [review]

**Review by robzor92 — APPROVED** [review]

**Review by robzor92 — APPROVED** [review]

**Review by javierdlrm — APPROVED** [review]

**Pull request merged**